pkg/kubelet/cm/deviceplugin/device_plugin_handler.go

```diff
@@ -0,0 +1,289 @@
+func NewHandlerImpl(updateCapacityFunc func(v1.ResourceList)) (*HandlerImpl, error) {
+	glog.V(2).Infof("Creating Device Plugin Handler")
+	handler := &HandlerImpl{
+		allDevices:       make(map[string]sets.String),
+		allocatedDevices: make(map[string]sets.String),
+		podDevices:       make(podDevices),
+	}
+
+	deviceManagerMonitorCallback := func(resourceName string, added, updated, deleted []pluginapi.Device) {
+		var capacity = v1.ResourceList{}
+		kept := append(updated, added...)
+		if _, ok := handler.allDevices[resourceName]; !ok {
+			handler.allDevices[resourceName] = sets.NewString()
+		}
+		// For now, Handler only keeps track of healthy devices.
+		// We can revisit this later when the need comes to track unhealthy devices here.
+		for _, dev := range kept {
+			if dev.Health == pluginapi.Healthy {
+				handler.allDevices[resourceName].Insert(dev.ID)
+			} else {
+				handler.allDevices[resourceName].Delete(dev.ID)
+			}
+		}
+		for _, dev := range deleted {
+			handler.allDevices[resourceName].Delete(dev.ID)
+		}
+		capacity[v1.ResourceName(resourceName)] = *resource.NewQuantity(int64(handler.allDevices[resourceName].Len()), resource.DecimalSI)
+		updateCapacityFunc(capacity)
+	}
+
+	mgr, err := NewManagerImpl(pluginapi.KubeletSocket, deviceManagerMonitorCallback)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to initialize device plugin manager: %+v", err)
+	}
+
+	handler.devicePluginManager = mgr
+	handler.devicePluginManagerMonitorCallback = deviceManagerMonitorCallback
+	// Loads in allocatedDevices information from disk.
+	err = handler.readCheckpoint()
+	if err != nil {
+		glog.Warningf("Continue after failing to read checkpoint file. Device allocation info may NOT be up-to-date. Err: %v", err)
+	}
+	return handler, nil
+}
+
+// Start starts device plugin registration service.
+func (h *HandlerImpl) Start() error {
+	return h.devicePluginManager.Start()
+}
+
+// Devices returns all of registered devices keyed by resourceName.
+func (h *HandlerImpl) Devices() map[string][]pluginapi.Device {
+	return h.devicePluginManager.Devices()
+}
+
+// Returns list of device Ids we need to allocate with Allocate rpc call.
+// Returns empty list in case we don't need to issue the Allocate rpc call.
+func (h *HandlerImpl) devicesToAllocate(podUID, contName, resource string, required int) (sets.String, error) {
+	h.Lock()
+	defer h.Unlock()
+	needed := required
+	// Gets list of devices that have already been allocated.
+	// This can happen if a container restarts for example.
+	devices := h.podDevices.containerDevices(podUID, contName, resource).Union(sets.NewString())
+	if devices != nil {
+		glog.V(3).Infof("Found pre-allocated devices for resource %s container %q in Pod %q: %v", resource, contName, podUID, devices.List())
+		needed = needed - devices.Len()
+	}
+	if needed == 0 {
+		// No change, no work.
+		return nil, nil
+	}
+	if needed < 0 {
+		// Pod is downsizing its resource requests. We need to make Allocate rpc call
```

Comment:
A pod's resource cannot be updated once admitted by the API server. I'd much rather fail loudly in this case (not crashloop) than handle it.